agent/php_txn.c

```diff
@@ -1109,3 +1109,92 @@ nr_status_t nr_php_txn_end(int ignoretxn, int in_post_deactivate TSRMLS_DC) {
 
   return NR_SUCCESS;
 }
+
+extern void nr_php_txn_add_code_level_metrics(
+    nr_attributes_t* attributes,
+    const nr_php_execute_metadata_t* metadata) {
+/* Current CLM functionality only works with PHP 7+ */
+#if ZEND_MODULE_API_NO < ZEND_7_0_X_API_NO /* PHP < PHP7 */
+  (void)attributes;
+  (void)metadata;
+  return;
+}
+#else
+  /*
+   * Check if code level metrics are enabled in the ini.
+   * If they aren't, exit and don't add any attributes.
+   */
+  if (!NRINI(code_level_metrics_enabled)) {
+    return;
+  }
+
+  if (NULL == metadata) {
+    return;
+  }
+
+  /*
+   * At a minimum, at least one of the following attribute combinations MUST be
+   * implemented in order for customers to be able to accurately identify their
+   * instrumented functions:
+   *  - code.filepath AND code.function
+   *  - code.namespace AND code.function
+   *
+   * If we don't have the minimum requirements, exit and don't add any
+   * attributes.
+   *
+   * Additionally, none of the needed attributes can exceed 255 characters.
+   */
+
+  const char* namespace = metadata->scope ? ZSTR_VAL(metadata->scope) : NULL;
+  const char* function
+      = metadata->function ? ZSTR_VAL(metadata->function) : NULL;
+  const char* filepath
+      = metadata->filepath ? ZSTR_VAL(metadata->filepath) : NULL;
+
+#define CHK_CLM_STRLEN(s)                         \
+  if (CLM_STRLEN_MAX < NRSAFELEN(nr_strlen(s))) { \
+    s = NULL;                                     \
+  }
+
+  CHK_CLM_STRLEN(filepath)
+  CHK_CLM_STRLEN(namespace)
+  CHK_CLM_STRLEN(function)
+
```

Comment:
Since these strings are originated as Zend values do we not already have the length from the Zend engine?  Might save some time.

Reply:
Please see [8c2a251](https://github.com/newrelic/newrelic-php-agent/pull/585/commits/8c2a251d0ac61bb3b9868ce1297f0095540aa0d3)

Reply:
Looks good!